sql/core/src/main/scala/org/apache/spark/sql/execution/python/MapInPandasExec.scala

```diff
@@ -28,7 +28,8 @@ import org.apache.spark.sql.execution.SparkPlan
 case class MapInPandasExec(
     func: Expression,
     output: Seq[Attribute],
-    child: SparkPlan)
+    child: SparkPlan,
+    override val isBarrier: Boolean)
```

Comment:
The feature `mapInPandas / mapInArrow` supporting barrier mode is only for ML use case, we don't need to complicate it. We only need to support such scenario:

```
input_dataset.mapInPandas(..., is_barrier=True).collect()
```

Reply:
I meant that the optimizers can might switch the physical plan order in the future, and that's why the implementation is flaky. Now this works because we don't switch the order anywhere else.

> Can we modify optimizer code to prevent it change the plan when it finds "is_barrier" set ?

To have a complete implementation, we should have something like https://github.com/apache/spark/pull/19873 which brought a lot of side effect, and it got reverted in the end.

Reply:
> If this is only for internal purpose, let's probably don't expose is_barrier: bool = False to the API surface at least.

No. It should be public API. Third-party lib such as xgboost also need to use it.